### ISSUE TITLE
[Key Vault] Verify pagination tests cover more than one page

### ISSUE THREAD

**Issue by timovv** [Azure/azure-sdk-for-net#30662]
Raising this issue following up from last week's team meeting.

As part of this month's release cycle, I released a migration of the Key Vault packages to JavaScript's Core v2 HTTP pipeline implementation. This change required some changes to the pagination methods each of the SDKs. One method in keyvault-secrets wasn't migrated properly, leading to a regression where the pagination would fail, but only from the second page onwards (the first page was fetched just fine).

This regression wasn't caught because, as it turns out, our pagination tests didn't actually paginate: the data returned from the service in the test wasn't enough to trigger pagination by default, and no maximum page size was set. This meant that fetching the next page wasn't covered in our tests and the issue slipped through.

I've since released a hotfix for JS fixing the issue and the tests (setting a maximum page size of 1 to force pagination), but we should ensure the other languages don't have this gap in test coverage.

Things to check:
* Make sure that the pagination tests for your language cover the scenario when multiple pages are returned
* If the scenario isn't covered, update the tests so that pagination is being tested properly (by e.g. setting a maximum page size).

See the [issue] and [resulting PR] from JavaScript for more background and the fix.

[issue]: https://github.com/Azure/azure-sdk-for-js/issues/22890
[resulting pr]: https://github.com/Azure/azure-sdk-for-js/pull/22898

**Comment by heaths** [comment]
Pagination is tested in Azure.Core, which we use for pagination but I can see if we make sure in Key Vault.

> If the scenario isn't covered, update the tests so that pagination is being tested properly (by e.g. setting a maximum page size).

@timovv I recommend mocking this. The page size is only a hint and can be ignored, and you end up having to create a bunch of secrets just to test multiple fetches, which all our SDKs' test frameworks should support mocking anyway. This isn't necessarily something to leave to chance if you know you had problems in the past.

**Comment by timovv** [comment]
Yeah, the bulk of the pagination logic is in Core for JS too, but the issue was that the convenience layer glue was calling the wrong method. I believe Key Vault, at least for JS, is a special case where the pagination is a bit more 'manual' than other SDKs so would have benefitted from the extra test coverage.

I will look into the mocking idea as you suggest -- I wasn't aware page size is just a hint! In the meantime, our test recordings now have multiple pages so it's being mocked at that level in a way.

**Comment by heaths** [comment]
Probably irrelevant for testing, but of related note: a page could still have 0 results but a `_nextLink` URI.

**Comment by pallavit** [comment]
@heaths can this issue be closed now?

**Comment by heaths** [comment]
I need to actually do the work, which I'm doing right now. We had coverage for secrets, but I added more to keys and certs and turns out certs wasn't setting the diagnostic scope. 🫢 PR forthcoming.
